### ISSUE TITLE
Multilingual Pretraining, Cross-Lingual Finetuning, and LID Heads with Zipformer

### ISSUE THREAD

**Issue by rohithkodali** [k2-fsa/icefall#2052]
I am currently exploring `pruned_transducer_stateless7` (Zipformer) and would like to understand the recommended workflows for transfer learning and multi-task setups in Icefall. I am coming from a NeMo/Conformer background and am looking for the equivalent recipes here.

I have three specific scenarios I need guidance on:

### 1. Cross-Lingual Finetuning (Language A -> Language B)
I want to take a Zipformer encoder pretrained on a resource-rich language (e.g., English/Librispeech) and finetune it on a low-resource language (e.g., Tamil).
* **Question:** What is the canonical way to handle the vocabulary mismatch?
* Is there an existing recipe that demonstrates freezing the Encoder and **re-initializing the Decoder/Joiner** for a new tokenizer/BPE? (I triained to freeze encoder and train with extended vocabulary of same lanaguge but accuracy is not better at all even on Librispeech) 
* Alternatively, is there a recipe for training **Language-Specific Adapters** while keeping the base Zipformer frozen?

Is this the right script to use  **/librispeech/SSL/zipformer**

### 2. Multilingual Pretraining
I am looking to train a single Zipformer model on a mix of languages.
* **Question:** What is the standard practice for input formatting? Do we simply concatenate the CutSets and rely on the model to learn, or is it recommended to prepend `<lang_id>` tokens to the prompt/supervision?
* Are there any reference recipes (e.g., creating a `MultiCutSet` logic) that you recommend following?

### 3. Joint ASR + Language Identification (LID)
I would like to add an auxiliary **LID Head** to the architecture to perform joint ASR and Language Identification.
* **Question:** Is there a supported way to attach a classification head to the Zipformer encoder outputs (e.g., via pooling `encoder_out`)?
* If no official recipe exists, could you point me to the right place in `zipformer.py` to hook in a secondary loss function for LID without disrupting the Pruned RNN-T loss?

Any pointers to specific `egs`, experimental branches, or config snippets would be greatly appreciated!

Thanks!

**Comment by KarelVesely84** [comment]
Hello Rohit,
i may have answers to some of your questions, although I am not part of the "core" team.

The `pruned_transducer_stateless7` is now outdated, the current model version is in [`zipformer`](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/zipformer) 
(or +/- 2 years ago it was the case).

**Concerning 1.)**
I tried a similar experiment recently (pre-trained model on 2k English, fine-tuned with 50hrs od Czech).
Inspired by [`finetune.py`](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/zipformer/finetune.py) I created a train tool that imports encoder parameters and randomly initializes predictor and joiner. It was important to set the batch counter for the `encoder` and `encoder_embed` modules to a high value (100k), so the dropouts a layer-drops behave ok (at the beginning of the training layer-drops act as if the network was more shalow).
However, in the end I was getting identical WER both for pre-trained and from-scratch training.

So it seems that pre-training for Zipformer is not as important as for other architectures (there are some tricks used to stabilize the training, decorelate internally the neuron signals and to keep the activations in healthy intervals).

**Concerning 2.)** 
i did not try that. There is an example of Chinese-English bilingual system:
https://github.com/k2-fsa/icefall/tree/master/egs/multi_zh_en/ASR
Prompting the transducer model is not possible (the context is fixed to 2 previous output tokens in predictor), 
prompting is suitable to transformer decoders. Also I've heard of masking the joiner outputs to fix to a set of a specific launguage, but there will be no code for that in icefall.

**Concerning 3.)**
I did not try that either. I did not see any "official hook", but the model is well structured internally, see the class [AsrModel](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/zipformer/model.py#L31)

Best regards
Karel

**Comment by rohithkodali** [comment]
Thanks @KarelVesely84 for the info.

**Comment by yfyeung** [comment]
Hi @KarelVesely84 @rohithkodali If your scenario involves only a small amount of supervised data (e.g., around 50 hours), you might consider collecting large-scale unlabeled data, by downloading audio from YouTube, and then trying our latest VietASR repository.

Repository: https://github.com/zzasdf/VietASR

Paper: https://arxiv.org/pdf/2505.21527

The approach is designed to leverage unlabeled data effectively and may be particularly suitable in low-resource supervised settings.

**Comment by asterixvn** [comment]
> Hello Rohit, i may have answers to some of your questions, although I am not part of the "core" team.
> 
> The `pruned_transducer_stateless7` is now outdated, the current model version is in [`zipformer`](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/zipformer) (or +/- 2 years ago it was the case).
> 
> **Concerning 1.)** I tried a similar experiment recently (pre-trained model on 2k English, fine-tuned with 50hrs od Czech). Inspired by [`finetune.py`](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/zipformer/finetune.py) I created a train tool that imports encoder parameters and randomly initializes predictor and joiner. It was important to set the batch counter for the `encoder` and `encoder_embed` modules to a high value (100k), so the dropouts a layer-drops behave ok (at the beginning of the training layer-drops act as if the network was more shalow). However, in the end I was getting identical WER both for pre-trained and from-scratch training.

Hi Karel,

I am working on finetuning of a multidomain RNN-T (zipformer) to a new domain (few hours). By using the original finetune.py which you mentioned above, I obtained good results. But I would like to change the tokenizer/BPE to a new task specific tokenizer. Could you please share your tool?

Thanks a lot.
Bac

**Comment by KarelVesely84** [comment]
> Hi Karel,
> 
> I am working on finetuning of a multidomain RNN-T (zipformer) to a new domain (few hours). By using the original finetune.py which you mentioned above, I obtained good results. But I would like to change the tokenizer/BPE to a new task specific tokenizer. Could you please share your tool?
> 
> Thanks a lot. Bac

Hello @asterixvn,
you can try this training script from my clone of icefall:
`https://github.com/KarelVesely84/icefall/blob/finetune_encoder/egs/librispeech/ASR/zipformer/finetune_encoder.py`

The CLI is a bit different from the default train.py scripts.
So far I did not test it with very small amounts of fine-tuning data.

Best regards,
Karel

**Comment by asterixvn** [comment]
Thanks a lot @KarelVesely84. It works with a smaller base_lr. However, in my case of domain adaptation, it's better to keep the original BPE and perform classical fine-tuning.

Best regards,
Bac
